pkg/sql/schemachanger/scbuild/builder_state.go

```diff
@@ -536,6 +536,25 @@ func (b *builderState) ComputedColumnExpression(tbl *scpb.Table, d *tree.ColumnT
 	return parsedExpr
 }
 
+// PartialIndexPredicateExpression implements the scbuildstmt.TableHelpers interface.
+func (b *builderState) PartialIndexPredicateExpression(
+	tableID catid.DescID, expr tree.Expr,
+) *scpb.Expression {
+	_, _, ns := scpb.FindNamespace(b.QueryByID(tableID))
+	tn := tree.MakeTableNameFromPrefix(b.NamePrefix(ns), tree.Name(ns.Name))
+	// TODO(fqazi): this doesn't work when referencing newly added columns.
+	_, err := schemaexpr.ValidatePartialIndexPredicate(
+		b.ctx,
+		b.descCache[tableID].desc.(catalog.TableDescriptor),
+		expr,
+		&tn,
+		b.semaCtx)
+	if err != nil {
+		panic(err)
+	}
```

Comment:
I encountered exactly the same issue when implementing `add constraint check` where I also need to validate the check expression in the builder. What I did is to refactor the code  in `schemaexpr.DequalifyAndValidateExpr` so that the caller will need to provide the columns, so I can circumvent this limitation.

You will benefit from that refactoring as well but this PR does not need to be blocked by it.